yogstation/code/modules/mining/lavaland/bloodbook.dm

```diff
@@ -0,0 +1,929 @@
+			animate(B, pixel_y = 30)
+			B.SpinAnimation(0.15 SECONDS)
+			phase++
+			addtimer(CALLBACK(src, PROC_REF(concavehead), user, target, phase, rend, B), 0.15 SECONDS)
+			animate(B,  pixel_y = 0, time = 0.2 SECONDS, easing = LINEAR_EASING)
+			return
+		if(3)
+			target.visible_message(span_warning("The meteor falls onto [target]'s head!"))
+			playsound(target,'sound/effects/meteorimpact.ogg', 10, 1)
+			for(var/mob/living/L in view(9, target))
+				shake_camera(L, 1, 1)
+			qdel(rend)
+			qdel(bowlingball)
+			splosion(user, target)
+	if(target)
+		sceneend(target, user)
+
+//PROPS
+
+/obj/structure/bed/killbubble
+	name = "crimson bubble"
+	desc = "A blood-red prison."
+	density = FALSE
+	anchored = TRUE
+	layer = MASSIVE_OBJ_LAYER
+	icon = 'icons/obj/projectiles.dmi'
+	icon_state = "leaper"
+	max_integrity = 500
+
+/obj/structure/bed/killbubble/proc/shred(var/times = 0)
+	if(times > 14)
+		return
+	var/obj/structure/slash/P = new(src.loc)
+	times++
+	addtimer(CALLBACK(src, PROC_REF(shred), times), 0.1 SECONDS)
+	addtimer(CALLBACK(GLOBAL_PROC, PROC_REF(qdel), P), 1.5 SECONDS)
+	return
+
+
+/obj/structure/bed/killbubble/proc/vacuum(mob/living/user)
+	for(var/atom/movable/A in range(src, 5))
+		if(A == user)
+			continue
+		if((!A.anchored))
+			var/shove_dir = get_dir(A, src)
+			A.Move(get_step(A, shove_dir), shove_dir)
+
+/obj/structure/prop
+	anchored = TRUE
+	density = FALSE
+	layer = MASSIVE_OBJ_LAYER
+	max_integrity = 500
+
+/obj/structure/prop/killrock
+	name = "ruby spire"
+	desc = "An unrelenting claw."
+	anchored = TRUE
+	layer = MASSIVE_OBJ_LAYER
+	icon = 'icons/obj/cult_64x64.dmi'
+	icon_state = "bloodstone-enter1"
+	var/stage = 1
+
+/obj/structure/prop/killrock/proc/stageupdate(var/stagechange = 0)
+	stage = stage + stagechange
+	icon_state = "bloodstone-enter[stage]"
+	return
+
+/obj/structure/prop/killrock/proc/stagechange(var/phase = 0)
+	switch(phase)
+		if(0)
+			++phase
+			stageupdate(1)
+			addtimer(CALLBACK(src, PROC_REF(stagechange), phase), 0.1 SECONDS)
+		if(1)
+			++phase
+			stageupdate(1)
+			addtimer(CALLBACK(src, PROC_REF(stagechange), phase), 0.1 SECONDS)
+			return
+		if(2)
+			++phase
+			stageupdate(-1)
+			addtimer(CALLBACK(src, PROC_REF(stagechange), phase), 0.1 SECONDS)
+			return
+		if(3)
+			++phase
+			stageupdate(-1)
+			addtimer(CALLBACK(src, PROC_REF(stagechange), phase), 0.1 SECONDS)
+			return
+
+/obj/structure/prop/killhand
+	name = "malevolent hand"
+	icon = 'icons/obj/projectiles.dmi'
+	icon_state = "cursehand0"
+	color = "#ff0000"
+
+/obj/structure/prop/killight
+	name = "crimson light"
+	icon = 'icons/obj/projectiles.dmi'
+	icon_state = "red_laser"
+	light_range = 2
+	light_power = 4
+	light_color = "#ff2828"
+	max_integrity = 500
+
+/obj/structure/prop/killight/proc/checknextspace(mob/living/L)
+	var/turf/front = get_step(L.loc, (src.dir))
+	if(ismineralturf(front))
+		var/turf/closed/mineral/M = front
+		M.attempt_drill()
```

Comment:
what happens if attempt_drill doesn't work for whatever reason?

Reply:
double checked and it stops wherever it currently is and returns false, ending the animation